proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
+                                st = Structures_A.thread_state.BlockedOnReceive ep r_opt pl")
+               apply (clarsimp simp: when_def split: option.splits)
+               apply (intro conjI impI allI; clarsimp simp: isReceive_def)
+                apply (corressimp corres: restart_thread_if_no_fault_corres[simplified isNothing_def])
+                apply (fastforce simp: obj_at_def)
+               apply (rule corres_guard_imp)
+                 apply (rule corres_split[OF _ reply_unlink_tcb_corres])
+                    apply (rule corres_guard_imp)
+                      apply (rule restart_thread_if_no_fault_corres[simplified isNothing_def])
+                     apply simp
+                    apply simp
+                   apply fastforce
+                  apply (wpsimp wp: reply_unlink_tcb_valid_sched_action)
+                 apply wpsimp
+                apply (fastforce simp: vs_all_heap_simps pred_tcb_at_def obj_at_def)
+               apply clarsimp
+              apply (prop_tac "\<not> isReceive st'")
+               apply (case_tac st; clarsimp simp: isReceive_def)
+              apply (case_tac st
+                     ; clarsimp simp: isReceive_def
+                     ; (corressimp corres: restart_thread_if_no_fault_corres[simplified isNothing_def]
+                        , fastforce simp: obj_at_def))
+             apply (wpsimp wp: gts_wp)
+            apply (wpsimp wp: gts_wp')
+           apply (clarsimp simp: vs_all_heap_simps obj_at_def is_tcb_def)
+          apply clarsimp
+         apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+           apply (wpsimp wp: restart_thread_if_no_fault_tcb_sts_of_other
+                             reply_unlink_tcb_tcb_sts_of_other gts_wp)
+         apply (rule hoare_weaken_pre
+               , rule cancel_all_ipc_loop_body_reply_unlink_ts_pred_other
+               , fastforce)
+        apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+        apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_not_runnable'_other, simp)
+       apply (clarsimp simp: pred_conj_def)
+       apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+       apply (wpsimp wp: cancel_all_ipc_loop_body_valid_sched)
+      apply (clarsimp simp: pred_conj_def)
+      apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+       apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_weak_sch_act_wf, simp)
+      apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_valid_queues, simp)
+     apply fastforce+
+  done
+
+crunches replyUnlink
+  for sch_act_not[wp]: "\<lambda>s. sch_act_not t s"
+  and typ_at'[wp]: "\<lambda>s. P (typ_at' T p s)"
+  (wp: crunch_wps)
+
+lemmas replyUnlink_typ_ats[wp] = typ_at_lifts[OF replyUnlink_typ_at']
+
+(* FIXME RT: is there an equivalent to typ_at_lifts? This fact doesn't seem to be produced
+             by reply_unlink_tcb_typ_at *)
```

Comment:
There is, yes: `abs_typ_at_lifts`

Reply:
(and it's better to use that than crunching it, so I'd recommend tweaking this one)

Reply:
Done, thanks!